control/server.py

```diff
@@ -330,43 +359,48 @@ def gateway_rpc_caller(self, requests, is_add_req):
         """Passes RPC requests to gateway service."""
         for key, val in requests.items():
             if key.startswith(GatewayState.BDEV_PREFIX):
-                if is_add_req:
-                    req = json_format.Parse(val, pb2.create_bdev_req())
-                    self.gateway_rpc.create_bdev(req)
-                else:
-                    req = json_format.Parse(val,
+                with self.rpc_lock:
```

Comment:
@baum the reason I placed the locks where I did is that we also want to protect the protobuf code. As when we get an exception it always in the protobuf functions.

Reply:
@gbregman please consider a simple gRPC service:

```python
class SimpleService(simple_proto_pb2_grpc.SimpleServiceServicer):
    def __init__(self):
        # Initialize shared resources
        self.shared_data = {}
        self.lock = threading.Lock()

    def simple_service_method(self, request, context):
        # Perform thread-safe operations
        with self.lock:
            # Access shared data
            value = self.shared_data.get(request.key)
            if value is not None:
                response = simple_proto_pb2.SimpleResponse(result=value)
            else:
                response = simple_proto_pb2.SimpleResponse(result="Not found")
        return response
```

In this example:

The SimpleService class is the gRPC service implementation. It uses a lock to protect access to shared resources (e.g., self.shared_data).

The simple_service_method method accesses shared data within a lock to ensure thread safety.
Additionally, gRPC itself is thread-safe for handling multiple requests and connections simultaneously. Thread safety for protocol buffer encoding and decoding is handled by the gRPC library itself, and you don't need to worry about it in your code.

Reply:
I recommend limiting the scope of this pull request to the following alterations:

1. Initialize the self.lock within the gRPC service's constructor.
2. In each gRPC method,  use ```with self.lock:```

Do you concur that the proposed approach effectively addresses the multithreading concern?
It seems no additional modifications are necessary unless I have overlooked something. Your feedback is appreciated.

Reply:
@baum I don't think it covers the entire issue. We wanted the pb2 stuff and the RPC function executed under the same lock. This might be an overkill but I'd rather do it like that. There are no strong reasons not to.

Reply:
Locking is only done in gRPC now, it seems that we don't need to lock the pb2 stuff.